File: src/2024/day1/1/main.go
package main

import (
	_ "embed"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

//go:embed data.txt
var fileContents string

func readAndProcessData(contents string) (int, error) {
	var firstColumn []int
	var secondColumn []int

	lines := strings.Split(contents, "\n")
	for _, line := range lines {
		// skip empty lines
		if line == "" {
			continue
		}

		// split the line and convert to integers
		parts := strings.Fields(line)
		if len(parts) < 2 {
			continue
		}

		first, err := strconv.Atoi(parts[0])
		if err != nil {
			return 0, err
		}
		second, err := strconv.Atoi(parts[1])
		if err != nil {
			return 0, err
		}

		firstColumn = append(firstColumn, first)
		secondColumn = append(secondColumn, second)
	}

	// sort columns
	sort.Ints(firstColumn)
	sort.Ints(secondColumn)

	// calculating
	var result int
	for i := range firstColumn {
		diff := firstColumn[i] - secondColumn[i]
		if diff < 0 {
			diff = -diff
		}
		result += diff
	}

	return result, nil
}

func main() {
	result, err := readAndProcessData(fileContents)
	if err != nil {
		fmt.Println("Error processing file:", err)
		return
	}

	fmt.Println(result)
}

File: src/2024/day1/1/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

fn read_and_process_data<P: AsRef<Path>>(filename: P) -> Result<i32, Box<dyn std::error::Error>> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);

    let mut first_column = Vec::new();
    let mut second_column = Vec::new();

    for line in reader.lines() {
        let line = line?;
        let nums: Vec<i32> = line
            .split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect();

        first_column.push(nums[0]);
        second_column.push(nums[1]);
    }

    first_column.sort_unstable();
    second_column.sort_unstable();

    // Calculate sum of absolute differences
    let result = first_column
        .iter()
        .zip(second_column.iter())
        .map(|(a, b)| (a - b).abs())
        .sum();

    Ok(result)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let root_dir = std::env::current_dir()?;
    let filename = root_dir.join("data.txt");

    match read_and_process_data(&filename) {
        Ok(result) => {
            println!("{}", result);
            Ok(())
        }
        Err(e) => {
            eprintln!("Error processing file: {}", e);
            Err(e)
        }
    }
}